aiida/tools/graph/age_rules.py

```diff
@@ -0,0 +1,430 @@
+
+    def __init__(self, max_iterations, track_edges):
+        """Initialization method
+
+        :param max_iterations: maximum number of iterations to perform.
+        :param bool track_edges: if True, will also track and return the edges traversed.
+        """
+        self.set_max_iterations(max_iterations)
+        self._track_edges = track_edges
+        self._iterations_done = None
+
+    def set_max_iterations(self, max_iterations):
+        """Sets the max iterations"""
+        valid_input = isinstance(max_iterations, int) or max_iterations is np.inf
+        if not valid_input:
+            raise TypeError('max_iterations has to be an integer or np.inf')
+        self._max_iterations = max_iterations
+
+    @property
+    def iterations_done(self):
+        """Number of iterations performed"""
+        return self._iterations_done
+
+    @abstractmethod
+    def run(self, operational_set):
+        """
+        Takes the operational_set and overwrites it with the set of nodes that results
+        from applying the rule (this might or not include the initial set of nodes as
+        well depending on the rule).
+
+        :type operational_set: :py:class:`aiida.tools.graph.age_entities.Basket`
+        :param operational_set: initital set of nodes to be overwritten by the rule.
+        """
+
+
+class QueryRule(Operation, metaclass=ABCMeta):
+    """Parent class for every rule that implements a query.
+
+    QueryRules take a generic QueryBuilder instance and a set of starting nodes and then
+    perform successive iterations of that query, each one from the set of nodes that the
+    previous one found. Depending on the class of rule used the final result will be
+    either the whole set of nodes traversed (UpdateRule), or only the final set of nodes
+    found in the last iteration of the query (ReplaceRule).
+    """
+
+    def __init__(self, querybuilder, max_iterations=1, track_edges=False):
+        """Initialization method
+
+        :param querybuilder: an instance of the QueryBuilder class from which to take the
+            procedure for traversal
+        :param int max_iterations: the number of iterations to run this query on (must be
+            a finite number for ReplaceRules)
+        :param bool track_edges: whether to track which edges are traversed and store them
+        """
+        super().__init__(max_iterations, track_edges=track_edges)
+
+        def get_spec_from_path(queryhelp, idx):
+            if (
+                queryhelp['path'][idx]['entity_type'].startswith('node') or
+                queryhelp['path'][idx]['entity_type'].startswith('data') or
+                queryhelp['path'][idx]['entity_type'].startswith('process') or
+                queryhelp['path'][idx]['entity_type'] == ''
+            ):
+                # not really nice, this way, how about using
+                # from aiida.orm.utils.node import is_valid_node_type_string and checking entity_type?
```

Comment:
Should this be answered?

Reply:
So, I don't think the method `is_valid_node_type_string` is currently checking the same as we are here. Since I am even less sure if these are compatible checks, I think I will leave this one as is for the moment. I will copy the comment in my "TO-DO" list to maybe resurface the possibility in an issue.